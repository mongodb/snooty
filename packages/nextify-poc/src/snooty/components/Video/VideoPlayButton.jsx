import React from 'react';
import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { palette } from '@leafygreen-ui/palette';
import LeafyButton from '@leafygreen-ui/button';
import PlayIcon from '@leafygreen-ui/icon/dist/Play';

const StyledButton = styled(LeafyButton)`
  color: ${palette.white};
  background-color: #0c1c27;
  border: 1px solid ${palette.gray.light3};
  border-radius: 50% !important;
  height: 80px;
  width: 80px;
  &:hover,
  &:active {
    border: none !important;
    box-shadow: none !important;
    color: ${palette.gray.dark2};
  }
`;

// Handles concentric circles on hover
const playButtonBorderStyling = css`
  border: 8px solid transparent;
  border-radius: 50% !important;
  box-shadow: 0px 0px 0px 8px transparent;
  :hover {
    background-color: ${palette.gray.light1 + 'AA'};
    box-shadow: 0px 0px 0px 8px ${palette.gray.dark2 + '80'};
    transition: box-shadow 0.6s, background-color 0.6s;
  }
`;

const VideoPlayButton = () => {
  return (
    <div css={playButtonBorderStyling}>
      <StyledButton>
        <PlayIcon size="xlarge" />
      </StyledButton>
    </div>
  );
};

export default VideoPlayButton;
