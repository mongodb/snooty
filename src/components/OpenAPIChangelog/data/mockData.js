export const mockChangelog = [
  {
    date: '2024-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'removed',
            changes: [
              {
                change: 'resource version 2023-01-01 was removed, latest available resource version is 2023-02-01',
                changeCode: 'resource-version-removed',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
              {
                change:
                  "for the 'query' request parameter 'clusterName', default value was changed from 'default' to 'cluster0'",
                changeCode: 'request-parameter-default-value-changed',
                backwardCompatible: false,
              },
              {
                change: "added the optional property 'links' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'deprecate',
            sunsetDate: '2024-02-01',
            changes: [
              {
                change: 'deprecated by resource version 2023-02-01 and marked for removal for 2024-02-01.',
                changeCode: 'resource-version-deprecated',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/accessList',
        httpMethod: 'POST',
        operationId: 'createProjectIpAccessList',
        tag: 'Project IP Access List',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change: "added the optional property 'name' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
              {
                change: "added the optional property 'comment' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-22',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change:
                  "added the optional property 'withDefaultAlertsSettings' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

export const mockDiff = [
  {
    path: '/api/atlas/v2/federationSettings/{federationSettingsId}/connectedOrgConfigs',
    httpMethod: 'GET',
    operationId: 'listConnectedOrgConfigs',
    tag: 'Federated Authentication',
    changeType: 'update',
    changes: [
      {
        change: "removed the non-success response with the status '400'.",
        changeCode: 'response-non-success-status-removed',
        backwardCompatible: false,
      },
      {
        change: "added the new optional 'query' request parameter 'envelope'.",
        changeCode: 'new-optional-request-parameter',
        backwardCompatible: true,
      },
    ],
  },
  {
    path: '/api/atlas/v2/groups/{groupId}/accessList',
    httpMethod: 'POST',
    operationId: 'createProjectIpAccessList',
    tag: 'Project IP Access List',
    changeType: 'removed',
    changes: [
      {
        change: "removed the required property 'name' from the response with the '200' status.",
        changeCode: 'response-required-property-removed',
        backwardCompatible: false,
      },
    ],
  },
];

export const mockIndex = {
  runDate: '2023-04-03',
  runId: '20230403105948111',
  specRevision: '422ab47f864909549362a7b39688404c82e2540b',
  specRevisionShort: '422ab47f864',
  versions: ['2023-01-01', '2023-02-01', '2023-08-11~preview'],
};
