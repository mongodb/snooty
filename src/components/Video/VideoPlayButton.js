import React from 'react';
import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { theme } from '../../theme/docsTheme.js';
import LeafyButton from '@leafygreen-ui/button';
import PlayIcon from '@leafygreen-ui/icon/dist/Play';

const playStyles = (theme) => css`
  background-color: #0c1c27;
  border: 1px solid #f9fbfa;
  border-radius: 50% !important;
  height: 80px;
  width: 80px;
  padding: ${theme.size.default};
  position: relative;

  :hover {
    background-color: #f9fbfa;
    border-color: #f9fbfa;
    color: #000000;
    transition: color 0.4s;
  }
`;

const StyledButton = styled(LeafyButton)`
  margin: -1px;
  &:hover,
  &:active {
    border: none !important;
    box-shadow: none !important;
  }
  border-radius: 50% !important;
  display: inline-block;
  padding: ${theme.size.default};
  position: relative;
  color: #f9fbfa;

  ${({ play, theme }) => play && playStyles(theme)}
`;

const playButtonBorderStyling = css`
  border: 8px solid transparent;
  border-radius: 50% !important;
  outline: 8px solid transparent;
  :hover {
    border: 8px solid rgba(159, 161, 162, 0.6);
    transition: border 0.6s;
    outline: 8px solid rgba(61, 79, 88, 0.6);
    transition: outline 0.2s;
  }
`;

const VideoPlayButton = ({ children, href, play, to, hasArrow = true, ...props }) => {
  return (
    <div css={playButtonBorderStyling}>
      <StyledButton play={play} {...props}>
        <PlayIcon size="xlarge" />
      </StyledButton>
    </div>
  );
};

export default VideoPlayButton;
