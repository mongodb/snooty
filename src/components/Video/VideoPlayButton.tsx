import React from 'react';
import styled from '@emotion/styled';
import { palette } from '@leafygreen-ui/palette';
import LeafyButton from '@leafygreen-ui/button';
import { css } from '@leafygreen-ui/emotion';
import Icon from '@leafygreen-ui/icon';

const StyledButton = styled(LeafyButton)`
  color: ${palette.white};
  background-color: #0c1c27;
  border: 1px solid ${palette.gray.light3};
  border-radius: 50% !important;
  height: 80px;
  width: 80px;
  &:hover,
  &:active {
    border: none !important;
    box-shadow: none !important;
    color: ${palette.gray.dark2};
  }
`;

// Handles concentric circles on hover
const playButtonBorderStyling = css`
  border: 8px solid transparent;
  border-radius: 50% !important;
  box-shadow: 0px 0px 0px 8px transparent;
  :hover {
    background-color: ${palette.gray.light1 + 'AA'};
    box-shadow: 0px 0px 0px 8px ${palette.gray.dark2 + '80'};
    transition: box-shadow 0.6s, background-color 0.6s;
  }
`;

const VideoPlayButton = () => {
  return (
    <div className={playButtonBorderStyling}>
      <StyledButton>
        <Icon size="xlarge" glyph="Play" />
      </StyledButton>
    </div>
  );
};

export default VideoPlayButton;
