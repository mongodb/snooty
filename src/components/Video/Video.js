import React from 'react';
import ReactPlayer from 'react-player';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { theme } from '../../theme/docsTheme';
import VideoPlayButton from './VideoPlayButton';

const ReactPlayerWrapper = styled('div')`
  margin-bottom: ${theme.size.articleContent};
  position: relative;
  padding-top: 56.25%;
`;

const StyledReactPlayer = styled(ReactPlayer)`
  position: absolute;
  top: 0;
  left: 0;
  > div {
    /* Redefines stacking context, allowing play button animation to stick on top */
    position: sticky;
  }
`;

const videoStyling = css`
  * {
    border-radius: 16px !important;
  }
`;

const Video = ({ nodeData: { argument }, ...rest }) => {
  return (
    <ReactPlayerWrapper>
      <StyledReactPlayer
        css={videoStyling}
        config={{
          youtube: {
            playerVars: {
              autohide: 1,
              modestbranding: 1,
              rel: 0,
            },
          },
        }}
        controls
        url={argument[0]['refuri']}
        width="100%"
        height="100%"
        playing
        playIcon={<VideoPlayButton />}
        light={<VideoPlayButton />}
      />
    </ReactPlayerWrapper>
  );
};

Video.propTypes = {
  nodeData: PropTypes.shape({
    argument: PropTypes.array.isRequired,
  }).isRequired,
};

export default Video;
