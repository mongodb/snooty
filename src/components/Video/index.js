import React, { useState, useEffect } from 'react';
import ReactPlayer from 'react-player/youtube';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/react';
import { withPrefix } from 'gatsby';
import { theme } from '../../theme/docsTheme';
import VideoPlayButton from './VideoPlayButton';

const ReactPlayerWrapper = styled('div')`
  position: relative;
  padding-top: 56.25%;
  margin: ${theme.size.medium} 0px;
`;

const StyledReactPlayer = styled(ReactPlayer)`
  position: absolute;
  top: 0;
  left: 0;
  > div {
    /* Redefines stacking context, allowing play button animation to stick on top */
    position: sticky;
  }
`;

const videoStyling = css`
  * {
    border-radius: 16px !important;
  }
`;

const Video = ({ nodeData: { argument }, ...rest }) => {
  const url = `${argument[0]['refuri']}`;
  const playable = ReactPlayer.canPlay(url);
  // use placeholder image for video thumbnail if invalid URL provided
  const [previewImage, setPreviewImage] = useState(withPrefix('assets/meta_generic.png'));

  useEffect(() => {
    // handles URL validity checking for well-formed YT links
    if (url.includes('youtube') || url.includes('youtu.be')) {
      const https = require('https');
      let testUrlValidity = 'https://www.youtube.com/oembed?url=' + url + '&format=json';

      https.get(testUrlValidity, (res) => {
        // if valid URL, display default YT thumbnail
        if (res.statusCode === 200) {
          setPreviewImage(true);
        }
      });
    }
  }, [url]);

  // handles remaining cases for invalid video URLs
  if (!playable) {
    console.warn(`Invalid URL: ${url} has been passed into the Video component`);
    return null;
  }

  return (
    <ReactPlayerWrapper>
      <StyledReactPlayer
        css={videoStyling}
        config={{
          youtube: {
            playerVars: {
              autohide: 1,
              modestbranding: 1,
              rel: 0,
            },
          },
        }}
        controls
        url={url}
        width="100%"
        height="100%"
        playing
        playIcon={<VideoPlayButton />}
        light={previewImage}
      />
    </ReactPlayerWrapper>
  );
};

Video.propTypes = {
  nodeData: PropTypes.shape({
    argument: PropTypes.array.isRequired,
  }).isRequired,
};

export default Video;
