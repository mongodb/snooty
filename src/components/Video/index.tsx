import React, { useState, useEffect, useMemo } from 'react';
import ReactPlayerYT from 'react-player/youtube';
import ReactPlayerWistia from 'react-player/wistia';
import styled from '@emotion/styled';
import { css } from '@leafygreen-ui/emotion';
import { palette } from '@leafygreen-ui/palette';
import { withPrefix } from 'gatsby';
import { theme } from '../../theme/docsTheme';
import { STRUCTURED_DATA_CLASSNAME, VideoObjectSd } from '../../utils/structured-data';
import type { VideoNode } from '../../types/ast';
import VideoPlayButton from './VideoPlayButton';

// Imported both players to keep bundle size low and rendering the one associated to the URL being passed in
const REACT_PLAYERS = {
  yt: {
    player: ReactPlayerYT,
    config: {
      youtube: {
        playerVars: {
          autohide: 1,
          modestbranding: 1,
          rel: 0,
        },
      },
    },
    name: 'youtube',
  },
  wistia: {
    player: ReactPlayerWistia,
    config: {},
    name: 'wistia',
  },
};

const ReactPlayerWrapper = styled('div')`
  position: relative;
  padding-top: 56.25%;
  margin: ${theme.size.medium} 0px;
  border-radius: 16px;
  overflow: hidden;
`;

const videoStyling = ({ name }: { name: string }) => css`
  position: absolute;
  top: 0;
  left: 0;
  > div {
    /* Redefines stacking context, allowing play button animation to stick on top */
    position: sticky;
  }

  ${name === 'wistia' &&
  `video {
    background: ${palette.white} !important
  }`}
`;

const getTheSupportedMedia = (url: string) => {
  let supportedType: 'yt' | 'wistia' | null = null;

  if (url.includes('youtube') || url.includes('youtu.be')) {
    supportedType = 'yt';
  }

  if (url.includes('wistia')) {
    supportedType = 'wistia';
  }

  return REACT_PLAYERS[supportedType as keyof typeof REACT_PLAYERS];
};

interface VideoProps {
  nodeData: VideoNode;
}

const Video = ({ nodeData: { argument, options } }: VideoProps) => {
  const url = `${argument[0]['refuri']}`;
  // use placeholder image for video thumbnail if invalid URL provided
  const [previewImage, setPreviewImage] = useState<string | boolean>(withPrefix('assets/meta_generic.png'));
  const { title, description, 'upload-date': uploadDate, 'thumbnail-url': thumbnailUrl } = options || {};
  const videoObjectSd = useMemo(() => {
    const sd = new VideoObjectSd({ embedUrl: url, name: title, uploadDate, thumbnailUrl, description });
    return sd.isValid() ? sd.toString() : undefined;
  }, [url, title, uploadDate, thumbnailUrl, description]);

  // handles URL validity checking for well-formed YT links
  const checkUrlValidity = async (url: string) => {
    if (url.includes('youtube') || url.includes('youtu.be')) {
      let testUrlValidity = 'https://www.youtube.com/oembed?url=' + url + '&format=json';

      const res = await fetch(testUrlValidity);
      if (res?.ok) {
        setPreviewImage(true);
      }
    }
  };

  useEffect(() => {
    checkUrlValidity(url);
  }, [url]);

  const ReactSupportedMedia = getTheSupportedMedia(url);

  if (!ReactSupportedMedia) {
    console.warn(`Media Not Found: A video player could not be found for the following ${url}`);
    return null;
  }

  const ReactPlayer = ReactSupportedMedia.player as any;
  const playable = (ReactPlayer as unknown as typeof ReactPlayer).canPlay?.(url);

  // handles remaining cases for invalid video URLs
  if (!playable) {
    console.warn(`Invalid URL: ${url} has been passed into the Video component`);
    return null;
  }

  return (
    <>
      {videoObjectSd && (
        <script
          className={STRUCTURED_DATA_CLASSNAME}
          type="application/ld+json"
          dangerouslySetInnerHTML={{ __html: videoObjectSd }}
        />
      )}
      <ReactPlayerWrapper>
        <ReactPlayer
          className={videoStyling(ReactSupportedMedia)}
          config={ReactSupportedMedia.config}
          controls
          url={url}
          width="100%"
          height="100%"
          playing
          playIcon={<VideoPlayButton />}
          light={previewImage}
        />
      </ReactPlayerWrapper>
    </>
  );
};

export default Video;
