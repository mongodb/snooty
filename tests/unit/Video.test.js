import React from 'react';
import { render } from '@testing-library/react';
import Video from '../../src/components/Video';

// data for this component
import mockData from './data/Video.test.json';

it('YouTube video renders correctly', () => {
  const tree = render(<Video nodeData={mockData.validYouTubeURL} />);
  expect(tree.asFragment()).toMatchSnapshot();
});

it('Vimeo video renders correctly', () => {
  const tree = render(<Video nodeData={mockData.validVimeoURL} />);
  expect(tree.asFragment()).toMatchSnapshot();
});

it('DailyMotion video renders correctly', () => {
  const tree = render(<Video nodeData={mockData.validDailyMotionURL} />);
  expect(tree.asFragment()).toMatchSnapshot();
});

describe('Console warning messages', () => {
  let consoleOutput = [];

  const mockedWarn = (output) => consoleOutput.push(output);
  beforeEach(() => (console.warn = mockedWarn));

  it('has to show in console warning messages', () => {
    render(<Video nodeData={mockData.invalidVideoURL1} />);
    render(<Video nodeData={mockData.invalidVideoURL2} />);
    expect(consoleOutput).toEqual([
      'Invalid URL: https://docs.mongodb.com/realm/ has been passed into the Video component',
      'Invalid URL: https://www.youtube.com/ has been passed into the Video component',
    ]);
  });
});
