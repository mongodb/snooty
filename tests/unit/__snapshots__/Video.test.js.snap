// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video renders correctly 1`] = `
<DocumentFragment>
  .emotion-2 {
  position: relative;
  padding-top: 56.25%;
}

.emotion-0 {
  position: absolute;
  top: 0;
  left: 0;
}

.emotion-0 > div {
  position: -webkit-sticky;
  position: sticky;
}

.emotion-0 * {
  border-radius: 16px !important;
}

<div
    class="emotion-2 emotion-3"
  >
    <div
      class="emotion-0 emotion-1"
      style="width: 100%; height: 100%;"
    />
  </div>
</DocumentFragment>
`;
